Add support for loading and connecting the Web Extension devtools background page.

xeenon opened the pull request:
#### 0b519b877420bd9b534b8ca4b8341952ab698208
<pre>
Add support for loading and connecting the Web Extension devtools background page.
<a href="https://webkit.org/b/246485">https://webkit.org/b/246485</a>
<a href="https://rdar.apple.com/problem/114823326">rdar://problem/114823326</a>

Reviewed by Jeff Miller.

Hook up loading the devtools_page when Web Inspector opens, or an extension is loaded
when Inspectors are already open during load, or private browsing is toggled.

* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasInspectorBackgroundPage):
(WebKit::WebExtension::inspectorBackgroundPagePath):
(WebKit::WebExtension::populateInspectorPropertiesIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::setHasAccessInPrivateBrowsing):
(WebKit::WebExtensionContext::inspectorBackgroundPageIdentifiers const):
(WebKit::WebExtensionContext::decidePolicyForNavigationAction):
(WebKit::WebExtensionContext::webViewWebContentProcessDidTerminate):
(WebKit::WebExtensionContext::inspectorBackgroundPageURL const):
(WebKit::WebExtensionContext::openInspectors const):
(WebKit::WebExtensionContext::loadedInspectors const):
(WebKit::WebExtensionContext::isInspectorBackgroundPage const):
(WebKit::WebExtensionContext::loadInspectorBackgroundPagesDuringLoad):
(WebKit::WebExtensionContext::unloadInspectorBackgroundPages):
(WebKit::WebExtensionContext::loadInspectorBackgroundPagesForPrivateBrowsing):
(WebKit::WebExtensionContext::unloadInspectorBackgroundPagesForPrivateBrowsing):
(WebKit::WebExtensionContext::loadInspectorBackgroundPage):
(WebKit::WebExtensionContext::unloadInspectorBackgroundPage):
(WebKit::WebExtensionContext::inspectorWillOpen):
(WebKit::WebExtensionContext::inspectorWillClose):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load):
(WebKit::WebExtensionController::unload):
(WebKit::WebExtensionController::unloadAll):
(WebKit::WebExtensionController::addProcessPool):
(WebKit::WebExtensionController::removeProcessPool):
(WebKit::WebExtensionController::addUserContentController):
(WebKit::WebExtensionController::removeUserContentController):
(WebKit::WebExtensionController::removeWebsiteDataStore):
(WebKit::WebExtensionController::cookiesDidChange):
(WebKit::WebExtensionController::extensionContext const):
(WebKit::WebExtensionController::extensions const):
(WebKit::WebExtensionController::addItemsToContextMenu):
(WebKit::WebExtensionController::didStartProvisionalLoadForFrame):
(WebKit::WebExtensionController::didCommitLoadForFrame):
(WebKit::WebExtensionController::didFinishLoadForFrame):
(WebKit::WebExtensionController::didFailLoadForFrame):
(WebKit::WebExtensionController::handleContentRuleListNotification):
(WebKit::WebExtensionController::purgeOldMatchedRules):
(WebKit::WebExtensionController::resourceLoadDidSendRequest):
(WebKit::WebExtensionController::resourceLoadDidPerformHTTPRedirection):
(WebKit::WebExtensionController::resourceLoadDidReceiveChallenge):
(WebKit::WebExtensionController::resourceLoadDidReceiveResponse):
(WebKit::WebExtensionController::resourceLoadDidCompleteWithError):
(WebKit::WebExtensionController::inspectorWillOpen):
(WebKit::WebExtensionController::inspectorWillClose):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::openLocalInspectorFrontend):
(WebKit::WebInspectorUIProxy::closeFrontendPageAndWindow):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsInspectedWindowCocoa.mm:
(WebKit::WebExtensionAPIDevToolsInspectedWindow::tabId):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIDevToolsInspectedWindow.h:
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::getOrCreate):
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIDevToolsInspectedWindow.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::tabIdentifier const):
(WebKit::WebExtensionContextProxy::addInspectorBackgroundPageIdentifier):
(WebKit::WebExtensionContextProxy::addInspectorBackgroundPage):
(WebKit::WebExtensionContextProxy::isInspectorBackgroundPage const):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDevTools.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSuspendAllMediaPlayback.mm:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab initWithWindow:extensionController:]):

Canonical link: <a href="https://commits.webkit.org/274506@main">https://commits.webkit.org/274506@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/faa96bab18273c19f37ea37a9f66d0ffd31d30f7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18245 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41800 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35166 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41572 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15574 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32856 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39840 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34035 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13350 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13314 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34975 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43078 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35656 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35304 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39119 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11618 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37358 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15684 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15347 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5144 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

jeffmapple commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2868,8 +2906,7 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 {
-    // FIXME: <https://webkit.org/b/246485> Handle inspector background pages in the assert.
-    ASSERT(webView == m_backgroundWebView);
+    ASSERT(webView == m_backgroundWebView || isInspectorBackgroundPage(webView));
```
You need to conditionalize the call to `isInspectorBackgroundPage()` on `ENABLE(INSPECTOR_EXTENSIONS)`.

jeffmapple commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2909,11 +2946,176 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    if (isInspectorBackgroundPage(webView)) {
+        [webView loadRequest:[NSURLRequest requestWithURL:inspectorBackgroundPageURL()]];
+        return;
+    }
```
You need to conditionalize this on `ENABLE(INSPECTOR_EXTENSIONS)`.

jeffmapple commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -245,6 +245,9 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
     NSString *generatedBackgroundContent();
 
+    bool hasInspectorBackgroundPage();
+    NSString *inspectorBackgroundPagePath();
```
Shouldn't these declarations (and the others you're adding in this header) be conditionalized on `ENABLE(INSPECTOR_EXTENSIONS)`? I see you've made some of this implementations unconditional, but I'm not sure that's the right thing to do.

xeenon replied:
These were intentional, so even on iOS we can check. This will likely we exposed on the API object too, and it does not make sense to allow checking it to only work on macOS. (We might just have it always return NO on iOS.)

jeffmapple commented on Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
```diff
@@ -56,6 +56,7 @@ namespace WebKit {
 
 class ContextMenuContextData;
 class WebExtensionContext;
+class WebInspectorUIProxy;
```
The forward class declaration can be conditionalized on `ENABLE(INSPECTOR_EXTENSIONS)`.

jeffmapple commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -185,6 +190,9 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     using UserContentControllerProxySet = WeakHashSet<WebUserContentControllerProxy>;
 
+    using InspectorTabVector = Vector<std::pair<Ref<WebInspectorUIProxy>, RefPtr<WebExtensionTab>>>;
+    using TabIdentifierWebViewPair = std::pair<WebExtensionTabIdentifier, RetainPtr<WKWebView>>;
```
These should be conditionalized on `ENABLE(INSPECTOR_EXTENSIONS)`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
```diff
@@ -55,9 +55,9 @@ class WebExtensionAPIStorageArea : public WebExtensionAPIObject, public JSWebExt
-    NSUInteger maxItems();
-    NSUInteger maxWriteOperationsPerHour();
-    NSUInteger maxWriteOperationsPerMinute();
+    double maxItems();
+    double maxWriteOperationsPerHour();
+    double maxWriteOperationsPerMinute();
```
`NSUInteger` started giving me an error for some reason.

jeffmapple commented on Source/WebKit/Shared/Extensions/WebExtensionConstants.h:
```diff
@@ -35,7 +35,7 @@ static constexpr size_t webExtensionDeclarativeNetRequestMaximumNumberOfEnabledR
 
 // MARK: Storage constants.
-static constexpr double webExtensionUnlimitedStorageQuotaBytes = DBL_MAX;
+static constexpr double webExtensionUnlimitedStorageQuotaBytes = std::numeric_limits<double>::max();
```
This seems unrelated to this patch, probably at least warrants a mention in the commit message.

xeenon replied:
Will mention. Was just a drive-by.

jeffmapple commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -49,7 +49,10 @@
 #if ENABLE(INSPECTOR_EXTENSIONS)
     if (name == "devtools"_s)
-        return objectForKey<NSDictionary>(extensionContext().manifest(), @"devtools_page");
+        return objectForKey<NSString>(extensionContext().manifest(), @"devtools_page") && (page.isInspectorPage() || extensionContext().isInspectorBackgroundPage(page));
```
Can we share "devtools_page" with the definition in `WebExtensionCocoa.mm`?

xeenon replied:
Not easily. We use less static strings like that is WebKit when compared to Safari.

jeffmapple commented on Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
```diff
@@ -43,6 +44,22 @@ void WebExtensionContextProxy::addFrameWithExtensionContent(WebFrame& frame)
+    if (m_popupPageMap.contains(page))
+        return std::get<std::optional<WebExtensionTabIdentifier>>(m_popupPageMap.get(page));
+
+    if (m_tabPageMap.contains(page))
```
Is it reasonable to assume that we're more likely to find a match in `m_tabPageMap` for the page and check for that first?

xeenon replied:
I waffled on this. But it is more likely it will be a popup hit, since that is slightly more common than opening an extension URL in a tab.

jeffmapple replied:
Ah, I didn't realize `m_tabPageMap` contained tabs showing extension content only. I agree with you.

jeffmapple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -546,9 +546,9 @@
+		72F21E9828C9216B007BF315 /* 100x100-red.jp2 in Copy Resources */ = {isa = PBXBuildFile; fileRef = 72F21E9728C9214C007BF315 /* 100x100-red.jp2 */; };
 		7498C3D72A94435F009A387E /* TestUTIUtilities.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 7498C3CF2A94435F009A387E /* TestUTIUtilities.cpp */; };
 		74DEF2252A946FD800E034A3 /* TestUTIRegistry.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 74DEF21D2A946FD700E034A3 /* TestUTIRegistry.cpp */; };
-		72F21E9828C9216B007BF315 /* 100x100-red.jp2 in Copy Resources */ = {isa = PBXBuildFile; fileRef = 72F21E9728C9214C007BF315 /* 100x100-red.jp2 */; };
```
I assume Xcode just re-arranged these lines for some reason.

xeenon replied:
Yeah, not sure what this was about.

jeffmapple commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSuspendAllMediaPlayback.mm:
```diff
@@ -26,6 +26,7 @@
 #import "config.h"
 
 #import "PlatformUtilities.h"
+#import "Test.h"
```
This change seems unrelated to the rest of your patch, is it necessary?

xeenon replied:
It was needed to fix the build with UnifiedSources rearranging with the new test being added.

jeffmapple reviewed (approved):
Since there are a lot of these, it might be worth mentioning in the commit message that you did a bunch of drive-by fixes to explicitly use Ref/RefPtr instead of auto.

xeenon committed:
Add support for loading and connecting the Web Extension devtools background page.
https://webkit.org/b/246485
rdar://problem/114823326

Reviewed by Jeff Miller.

Hook up loading the devtools_page when Web Inspector opens, or an extension is loaded
when Inspectors are already open during load, or private browsing is toggled.

* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasInspectorBackgroundPage):
(WebKit::WebExtension::inspectorBackgroundPagePath):
(WebKit::WebExtension::populateInspectorPropertiesIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::setHasAccessInPrivateBrowsing):
(WebKit::WebExtensionContext::inspectorBackgroundPageIdentifiers const):
(WebKit::WebExtensionContext::decidePolicyForNavigationAction):
(WebKit::WebExtensionContext::webViewWebContentProcessDidTerminate):
(WebKit::WebExtensionContext::inspectorBackgroundPageURL const):
(WebKit::WebExtensionContext::openInspectors const):
(WebKit::WebExtensionContext::loadedInspectors const):
(WebKit::WebExtensionContext::isInspectorBackgroundPage const):
(WebKit::WebExtensionContext::loadInspectorBackgroundPagesDuringLoad):
(WebKit::WebExtensionContext::unloadInspectorBackgroundPages):
(WebKit::WebExtensionContext::loadInspectorBackgroundPagesForPrivateBrowsing):
(WebKit::WebExtensionContext::unloadInspectorBackgroundPagesForPrivateBrowsing):
(WebKit::WebExtensionContext::loadInspectorBackgroundPage):
(WebKit::WebExtensionContext::unloadInspectorBackgroundPage):
(WebKit::WebExtensionContext::inspectorWillOpen):
(WebKit::WebExtensionContext::inspectorWillClose):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load):
(WebKit::WebExtensionController::unload):
(WebKit::WebExtensionController::unloadAll):
(WebKit::WebExtensionController::addProcessPool):
(WebKit::WebExtensionController::removeProcessPool):
(WebKit::WebExtensionController::addUserContentController):
(WebKit::WebExtensionController::removeUserContentController):
(WebKit::WebExtensionController::removeWebsiteDataStore):
(WebKit::WebExtensionController::cookiesDidChange):
(WebKit::WebExtensionController::extensionContext const):
(WebKit::WebExtensionController::extensions const):
(WebKit::WebExtensionController::addItemsToContextMenu):
(WebKit::WebExtensionController::didStartProvisionalLoadForFrame):
(WebKit::WebExtensionController::didCommitLoadForFrame):
(WebKit::WebExtensionController::didFinishLoadForFrame):
(WebKit::WebExtensionController::didFailLoadForFrame):
(WebKit::WebExtensionController::handleContentRuleListNotification):
(WebKit::WebExtensionController::purgeOldMatchedRules):
(WebKit::WebExtensionController::resourceLoadDidSendRequest):
(WebKit::WebExtensionController::resourceLoadDidPerformHTTPRedirection):
(WebKit::WebExtensionController::resourceLoadDidReceiveChallenge):
(WebKit::WebExtensionController::resourceLoadDidReceiveResponse):
(WebKit::WebExtensionController::resourceLoadDidCompleteWithError):
(WebKit::WebExtensionController::inspectorWillOpen):
(WebKit::WebExtensionController::inspectorWillClose):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::openLocalInspectorFrontend):
(WebKit::WebInspectorUIProxy::closeFrontendPageAndWindow):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDevToolsInspectedWindowCocoa.mm:
(WebKit::WebExtensionAPIDevToolsInspectedWindow::tabId):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIDevToolsInspectedWindow.h:
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::getOrCreate):
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIDevToolsInspectedWindow.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::tabIdentifier const):
(WebKit::WebExtensionContextProxy::addInspectorBackgroundPageIdentifier):
(WebKit::WebExtensionContextProxy::addInspectorBackgroundPage):
(WebKit::WebExtensionContextProxy::isInspectorBackgroundPage const):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDevTools.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSuspendAllMediaPlayback.mm:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab initWithWindow:extensionController:]):

Canonical link: https://commits.webkit.org/274506@main
